Updates to the API endpoints / docs, and fixing the Theme builder

abidlabs opened the pull request:
This WIP PR makes several changes to the API endpoints / docs:

* Allows one to pass in `False` to `api_name` to remove the endpoint from appearing in `/info` altogether
* Sets `api_name=False` to all of the event handlers in the theme builder so that the theme builder is able to launch
* Removes unnecessary boilerplate code from the JS code snippet and some other cleanup of the API docs page
* `gr.Examples` now takes in a parameter that can be used to set the api endpoint name, and this is set to `False` by default. This means that the the explicit named endpoint "load_examples" is no longer created. This was introduced in #4456 but I think it's a bad idea since it creates a second named endpoint in `gr.Interface`. This is technically a breaking change, however since it was just introduced, I think we can get away with it. 
* Makes some small formatting fixes to the client -- in part so that the client tests are triggered to make sure they all pass too 

Note that when testing the theme builder, you'll have to use your arrow keys to select the theme because the dropdown is not visible. That's being addressed in a separate PR: #4691 

Closes: #4565

abidlabs commented on client/python/gradio_client/client.py:
```diff
@@ -235,7 +235,8 @@ def duplicate(
             current_info.hardware or huggingface_hub.SpaceHardware.CPU_BASIC
         )
-        hardware = hardware or original_info.hardware
+        if hardware is None and original_info.hardware is not None:
```
Fixed a typing error here

freddyaboulton replied:
What's the type error? I thought these are equivalent.

<img width="224" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/7ab03dfe-f9cd-4ccf-8858-1603b9e9c013">

abidlabs replied:
Ok so it was the fact that pyright didn't like `hardware` getting reassigned to have a different type. Originally `hardware` could only be a string or null, but then after this step, it could also be `SpaceHardware`. Now, it can be `SpaceHardware` from the beginning, which might be useful.

abidlabs commented on client/python/gradio_client/client.py:
```diff
@@ -354,7 +355,7 @@ def view_api(
 
         Parameters:
-            all_endpoints: If True, prints information for both named and unnamed endpoints in the Gradio app. If False, will only print info about named endpoints. If None (default), will only print info about unnamed endpoints if there are no named endpoints.
+            all_endpoints: If True, prints information for both named and unnamed endpoints in the Gradio app. If False, will only print info about named endpoints. If None (default), will print info about named endpoints, unless there aren't any -- in which it will print info about unnamed endpoints.
```
Clearer language

freddyaboulton commented:
Is this ready for review @abidlabs ?

abidlabs commented:
Yes it is, thanks @freddyaboulton

freddyaboulton commented on gradio/blocks.py:
```diff
@@ -910,7 +914,7 @@ def render(self):
             for i, dependency in enumerate(self.dependencies):
                 api_name = dependency["api_name"]
-                if api_name is not None:
+                if api_name is not None and not False:
```
Is the intention to not run this code block if `api_name` is False? If so, I don't think this is working correctly. `not False` always evaluates to `True`:

<img width="411" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/5fd121b9-c3e0-44b0-a628-acfed308b2e4">

abidlabs replied:
Lol that should be: `if api_name is not None and api_name is not False`

freddyaboulton commented on guides/01_getting-started/03_sharing-your-app.md:
```diff
@@ -120,21 +120,21 @@ Note: if you use IFrames, you'll probably want to add a fixed `height` attribute
+![Use via API](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/gradio-guides/api3.gif)
 
-![Use via API](/assets/guides/use_via_api.png)
+This is a page that lists the endpoints that can be used to query the Gradio app, via our supported clients: either [the Python client](), or [the JavaScript client](). For each endpoint, Gradio automatically generates the parameters and their types, as well as example inputs.
```
nit - you should link to the client source or pypi/npm pages?

freddyaboulton commented on js/app/src/api_docs/CodeSnippets.svelte:
```diff
@@ -90,18 +90,14 @@ print(result)</pre>
 -->
-	const response_{i} = await fetch("{example_input}");
-	const example{component} = await response_{i}.blob();
+const response_{i} = await fetch("{example_input}");
```
This looks much nicer!

freddyaboulton commented on js/app/src/api_docs/CodeSnippets.svelte:
```diff
@@ -137,10 +133,7 @@ async function run() &lbrace;
-&rbrace;
-
-run();
+console.log(result.data);
```
Don't we need to keep `result?.data`

abidlabs replied:
I don't think it's necessary but @pngwn could you confirm?

abidlabs replied:
Will revert unless I hear otherwise from @pngwn

freddyaboulton reviewed (commented):
Looks good @abidlabs !

I think there is an issue where the python client will not be able to load a demo if any of the routes have `api_name=False`. We should probably add a test for this. Other than that, I think this looks about ready to merge.


<img width="889" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/9c2a7bcc-f277-4c8d-8237-a1dff18a89da">

abidlabs commented:
> I think there is an issue where the python client will not be able to load a demo if any of the routes have api_name=False. We should probably add a test for this. Other than that, I think this looks about ready to merge.

Thanks @freddyaboulton I'll take a look at what's happening here.

> What's the type error? I thought these are equivalent.

Let me double check I may have made a mistake here


--

Thanks for the detailed review @freddyaboulton!

abidlabs commented:
> I think there is an issue where the python client will not be able to load a demo if any of the routes have api_name=False. We should probably add a test for this. Other than that, I think this looks about ready to merge.

Fixed this and added a test

abidlabs commented:
Should be ready for a re-review @freddyaboulton!

freddyaboulton commented on client/python/test/test_client.py:
```diff
@@ -791,6 +782,12 @@ def test_unnamed_endpoints_use_fn_index(self, count_generator_demo):
+        # This demo has no api_name
+        with connect(count_generator_demo) as client:
+            info = client.view_api(return_format="dict")
+            assert len(info["named_endpoints"]) + len(info["unnamed_endpoints"]) == 2
```
Wouldn't `assert len(info["unnamed_endpoints"]) == 2 and len(info["named_endpoints"]) == 0 ` be better?

abidlabs replied:
Yea

abidlabs replied:
Don't know why I wrote it this way

freddyaboulton reviewed (approved):
Thanks for making the changes @abidlabs !

abidlabs commented:
Thanks @freddyaboulton for the great review! Will merge in after CI
